File: 42.trapping-rain-water.cs

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

public class Solution42 
{
    // Key: Use stack to persist the index if not being closed
    // Close if found a higher boundry. 
    // Be careful on corner cases.
    public int Trap(int[] height) {
        int result = 0;
        Stack<int> stack = new Stack<int>();
        for(int i = 0; i < height.Length; ++i)
        {
            var current = height[i];
            while(stack.Any() && current > height[stack.Peek()])
            {
                int top = stack.Pop();
                if(!stack.Any())
                    break;
                
                int distance = i - stack.Peek() - 1;
                int bound = Math.Min(current, height[stack.Peek()]) - height[top];

                result += distance * bound;
            }
            stack.Push(i);
        }

        return result;
    }
}

File: 42_trapping-rain-water.py
from typing import List


class Solution:
    def trap(self, height: List[int]) -> int:
        size = len(height)
        stack = []
        res = 0
        for i in range(size):
            cur = height[i]
            
            while len(stack) > 0 and cur > height[stack[-1]]:
                top = stack.pop()

                if len(stack) == 0: break

                distance = i - stack[-1] -1
                bound = min(cur, height[stack[-1]]) - height[top]

                res += distance* bound
            stack.append(i)
        
        return res

instance = Solution()
print(instance.trap([0,1,0,2,1,0,1,3,2,1,2,1]))